### ISSUE TITLE
code problems

### ISSUE THREAD

**Issue by SCIKings** [tensorly/tensorly#558]
In neural network compression, CP decomposition is often solved using the least squares method. is it necessary to divide the training set and test set for the weight tensor during CP decomposition? The following is the code I wrote from my understanding of CP decomposition, I don't know if it is correct. What else needs to be modified? I will be grateful if you can help me.

def khatri_rao_4(A, B, C):
    I, R = A.shape
    J, _ = B.shape
    K, _ = C.shape
    D = np.zeros((I * J * K, R))
    for r in range(R):
        D[:, r] = np.kron(np.kron(A[:, r], B[:, r]), C[:, r])
    return D

def unfold(tensor, mode):
    return np.reshape(np.moveaxis(tensor, mode, 0), (tensor.shape[mode], -1))

def cp_decomposition(tensor, R, max_iter=100, tol=1e-6):
    I, J, K, L = tensor.shape

    A = np.random.rand(I, R)
    B = np.random.rand(J, R)
    C = np.random.rand(K, R)
    D = np.random.rand(L, R)

    rec_errors=[]
    restruct_tensor=tensor
    for iteration in range(max_iter):
        X1 = unfold(tensor, 0)
        krc_bd = khatri_rao_4(D, C, B)
        A = np.linalg.solve(krc_bd.T @ krc_bd, krc_bd.T @ X1.T).T

        X2 = unfold(tensor, 1)
        krc_ad = khatri_rao_4(D, C, A)
        B = np.linalg.solve(krc_ad.T @ krc_ad, krc_ad.T @ X2.T ).T

        X3 = unfold(tensor, 2)
        krd_ab = khatri_rao_4(D, B, A)
        C = np.linalg.solve(krd_ab.T @ krd_ab, krd_ab.T @ X3.T).T

        X4 = unfold(tensor, 3)
        krb_ac = khatri_rao_4(A, C, B)
        D = np.linalg.solve(krb_ac.T @ krb_ac,  krb_ac.T @ X4.T).T

        restruct_tensor=np.einsum('ir,jr,kr,mr->ijkm', A, B, C, D)
        rec_error = np.linalg.norm(tensor - restruct_tensor)
        # print(rec_error)
        # print(iteration)
        rec_errors.append(rec_error)

        if iteration > 1:
            if abs(rec_errors[-2] - rec_errors[-1]) < tol:

                break

    return (A, B, C, D),restruct_tensor

if __name__ == "__main__":
    I, J, K, L = 50, 50, 50, 50
    tensor = np.random.rand(I, J, K, L)

    R = 5

    _,restruct = cp_decomposition(tensor, R)
    rank = 5  
    factors = parafac(tensor, rank=rank, init='random')[1]

    A,B,C,D = factors[0],factors[1],factors[2],factors[3]

    recon_tl = np.einsum('ir,jr,kr,mr->ijkm', A, B, C, D)
    error_tl = np.linalg.norm(restruct - recon_tl)

**Comment by aarmey** [comment]
Hi @SCIKings—I'm not quite sure what you are trying to accomplish here. The TensorLy Slack might be a better place to ask for advice.

https://github.com/tensorly/tensorly/discussions/302
